### ISSUE TITLE
[Idea] wasmtime dynamic library loading like with *.dll, *.so, *dylib in OS

### ISSUE THREAD

**Issue by redradist** [bytecodealliance/wasmtime#4194]
#### Feature

wasmtime dynamic library loading like with *.dll, *.so, *dylib in OS

#### Benefit

It will simplify loading a big application with it own dependecies

#### Implementation

It is implemented in https://github.com/bytecodealliance/wasmtime-cpp/issues/25#issuecomment-1132157937 as C++ class.
Lets consider the similar api for wasmtime that it would be available on all platforms

**Comment by redradist** [comment]
@alexcrichton Could you involve people from your team to this discussion ?

**Comment by alexcrichton** [comment]
My personal opinion on this is that this is probably a very ambitious issue and not the best place to discuss this. The issue description here is quite terse and doesn't convey details such as motivation, use cases, ecosystem interactions, etc. It's not 100% clear to me what you're shooting for so I'm assuming it's something akin to dynamic linking on native platforms where the general idea is that you can share a big was of code between programs by loading libraries at runtime. Designing such a feature for WebAssembly has large implications mostly on toolchains and how they work. For example while I believe Emscripten implements a form of dynamic linking I don't think it's been evaluated for off-the-web use yet.

Overall this would probably be best discussed in a larger forum that's not just the Wasmtime repository because the Wasmtime-specific pieces of this are likely to be quite simple compared to the toolchain and ecosystem changes necessary. Unfortunately I don't know the best forum in which to discuss this.

**Comment by redradist** [comment]
Yes it is about dynamic linking on systems like Linux, Windows and MacOS.
The reason is that it is not possible to merge all code in single wasm file.
The basic implement was done in simple class that I provided in wasmtime-cpp
